restic copy --stream: run one large copy operation crossing snapshot boundaries - issue #5453

wplapper opened the pull request:
restic copy now allows to optimize the repository packsize across multiple snapshots when copying from one repository to another repository.

changelog/unreleased/issue-5453:
announce new option `--batch`.

cmd/restic/cmd_check_integration_test.go:
add stream test

cmd/restic/cmd_copy.go:
new functions collectAllSnapshots(), copyTreeBatched()  to perfrom the batch copy.

cmd/restic/cmd_copy_integration_test.go:
a new streaming test

doc/045_working_with_repos.rst:
a short note about option `--batch`

<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
This PR modifies the way the ``copy`` operation works. With the ``--batch`` option, one large request list is built and sent to repository.Repack(), instead of one request list per snapshot to be copied. This allows optimizing the repository packsize while copying ``restic backup``with only minor changes between different snapshots. Normally, one new index file is created per snapshots, together with at least two packfiles for each new snapshot. When changes between backups are small, this can result in a lot of smallish packfiles. 

Yes, this can always be fixed by running ``restic prune``, but this is an extra step.  
<!--
Describe the changes and their purpose here, as detailed as needed.
-->

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
closes #5453
<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

- [X] I have added tests for all code changes.
- [X] I have added documentation for relevant changes (in the manual).
- [X] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [X] I'm done! This pull request is ready for review.

MichaelEischer reviewed (changes requested):
The `--stream` and non-stream code paths MUST be unified or this will become a maintenance nightmare.

MichaelEischer commented on cmd/restic/cmd_copy.go:
```diff
@@ -252,5 +380,54 @@ func copyTree(ctx context.Context, srcRepo restic.Repository, dstRepo restic.Rep
 	return nil
 }
+
+// printStats: generate short statistic for blob / packs / snapshots to be copied
```
Remove the stats function here and let's discuss it in https://github.com/restic/restic/pull/5464 .

wplapper commented on cmd/restic/cmd_copy.go:
```diff
@@ -120,9 +126,13 @@ func runCopy(ctx context.Context, opts CopyOptions, gopts GlobalOptions, args []
 		return ctx.Err()
 	}
 
+	if opts.streamAll {
```
reworked copy logic completely.

wplapper commented on cmd/restic/cmd_check_integration_test.go:
```diff
@@ -29,8 +29,8 @@ func testRunCheckOutput(gopts GlobalOptions, checkUnused bool) (string, error) {
-			ReadData:    true,
-			CheckUnused: checkUnused,
+			ReadData: true,
+			//CheckUnused: checkUnused,
```
done.

MichaelEischer commented on internal/repository/repack.go:
```diff
@@ -48,20 +48,33 @@ func Repack(
 }
 
+// RepackInner is a visible wrapper around repack()
+func RepackInner(ctx context.Context,
```
This is a horrible hack, but I get that `Repack` currently doesn't allow implementing this PR. So, let's do it properly instead.

Due to https://github.com/restic/restic/pull/5558 it's no longer possible to separately start the uploader and call flush. This is instead handled automatically via `WithBlobUploader`. 

But I don't see why we couldn't extract the `WithBlobUploader` call out of `Repack`. Then the call has to pass in the `restic.BlobSaver` but that is fine IMHO. I'd then also rename `Repack` to `CopyBlobs`.

MichaelEischer commented on cmd/restic/cmd_copy.go:
```diff
@@ -203,10 +180,13 @@ func copyTree(ctx context.Context, srcRepo restic.Repository, dstRepo restic.Rep
-		copyBlobs.Insert(h)
-		for _, p := range pb {
-			packList.Insert(p.PackID)
+		if !seenBlobs.Has(h.ID) {
```
I don't think we need `seenBlobs` at all. Once `repository.Repack(Inner)` has run, all copied blobs will immediately also be returned by `dstRepo.LookupBlobSize`, even if the repository was not flushed yet.

MichaelEischer replied:
Well, that wasn't actually the case and will be fixed by https://github.com/restic/restic/pull/5614 .

MichaelEischer commented on cmd/restic/cmd_copy.go:
```diff
@@ -243,10 +223,110 @@ func copyTree(ctx context.Context, srcRepo restic.Repository, dstRepo restic.Rep
+	var wgCtx context.Context
+	var seenBlobs restic.IDSet
+
+	if opts.batch {
```
With `repository.RepackInner` (or it's replacement) we can actually let copy automatically do the right thing. Now, that `copyTree` no longer flushes automatically, `copyTreeBatched` can just check that a reasonable amount of data has already been copied before flushing the repository. This should only add a minimal amount of extra code but has the benefit of offering optimal performance out of the box. 

In case you're wondering how to implement this flexible flush control with `WithBlobUploader`, take a look at the double loop at https://github.com/restic/restic/blob/c854338ad152098928c49136a2a2149db8243224/internal/repository/repack_test.go#L21 .

MichaelEischer reviewed (changes requested):
This will require further rework to be compatible with the changes from https://github.com/restic/restic/pull/5558 . While looking closer at `copyTreeBatched` I've noticed that it's actually possible to let it automatically batch multiple snapshots together. That should be possible with limited extra complexity and would have to benefit of just working out of the box.

wplapper committed:
restic copy --batch: a fresh start from commit 382616747

Instead of rebasing my code, I decided to start fresh, since WithBlobUploader()
has been introduced.

changelog/unreleased/issue-5453:
doc/045_working_with_repos.rst:
the usual

cmd/restic/cmd_copy.go:
gather all snaps to be collected - collectAllSnapshots()
run overall copy step - func copyTreeBatched()
helper copySaveSnapshot() to save the corresponding snapshot

internal/repository/repack.go:
introduce wrapper CopyBlobs(), which passes parameter `uploader restic.BlobSaver` from
WithBlobUploader() via copyTreeBatched() to repack().

internal/backend/local/local_windows.go:
I did not touch it, but gofmt did: whitespace

wplapper committed:
restic copy --batch - fussy linter

internal/repository/repack.go: I have to please the mighty linter.

wplapper committed:
restic copy --batch: The mighty linter

I cave in - no double comment

wplapper commented:
In principle, one could make ``--batch`` the default option and have ``--no-batch`` as a backup for some unusual use cases.

MichaelEischer committed:
repository: replace CopyBlobs with Repack implementation

MichaelEischer committed:
copy: remove bugous seenBlobs set

MichaelEischer committed:
don't sort snapshots, drop duplicate code and cleanup copyTreeBatched function signature

MichaelEischer committed:
use correct context

MichaelEischer committed:
repository: fix LookupBlobSize to also report pending blobs

MichaelEischer committed:
unify snapshot copy codepaths

MichaelEischer committed:
deduplicate blob enqueuing

MichaelEischer reviewed (changes requested):
Looks like I should have replied here a bit earlier. Adding a `--all` option is completely the opposite direction of what this PR should move into, sorry. I actually think that even a `--batch` option is already slightly too much. In particular, there's an apparent way that let's `copy` perform well even without additional configuration.

I might partially change my mind if a relevant performance benefit or other benefit can be demonstrated, but this will be an uphill battle.

As I want to merge this PR before I introduce even more merge conflicts (the next 2 or 3 are already prepared), I'll take over cleaning up the code here (including my comments).

MichaelEischer committed:
automatically batch snapshots in copy

MichaelEischer committed:
align docs

MichaelEischer committed:
Simplify test

MichaelEischer committed:
merge into existing copy test

wplapper commented on cmd/restic/cmd_copy.go:
```diff
@@ -268,3 +302,81 @@ func copyStats(srcRepo restic.Repository, copyBlobs restic.BlobSet, packList res
+	visitedTrees restic.IDSet, selectedSnapshots []*data.Snapshot, opts CopyOptions,
+	printer progress.Printer) error {
+
+	// seenBlobs is necessary in about 1 of 10000 blobs, in the other 99.99% the check
```
Sorry, I trusted your statement about the pending blobs. I tried to investigate the issue (since I knew that ``seenBlobs``was a circumvention), but I could not see an an apparent pattern. So - in order to deliver a correct result - I used ``seenBlobs``. I never suspected that the master index was the culprit.

wplapper commented:
Michael, thank you for cleaning up my messes :)

MichaelEischer committed:
more aggressive batching

MichaelEischer commented:
> In total about 110 more packfiles, and many more index files. This can easily be remedied with a `restic prune`, but this is an extra step.

Thanks for testing! Looks like I've effectively broken the PR for the snapshots you've tested with :see_no_evil: . The latest commit makes the batching quite a bit more aggressive, which should be enough to yield a single digit number of indexes and get rid of most of the difference in the pack files number. Can you give that another try?

MichaelEischer committed:
output which source and target snapshot belong together

MichaelEischer reviewed (approved):
LGTM. Thanks!
